crates/llms/tests/llms/mod.rs

```diff
@@ -0,0 +1,121 @@
+    /// This is used in snapshot testing to assert certain properties of the response.
+    pub json_path: Vec<(&'static str, &'static str)>,
+}
+/// Creates [`TestCase`] instances from request/response that JSON serialize to
+/// [`CreateChatCompletionRequest`] and [`CreateChatCompletionResponse`].
+#[macro_export]
+macro_rules! test_case {
+    ($name:expr, $req:expr, $jsonpaths:expr) => {
+        TestCase {
+            name: $name,
+            req: serde_json::from_value($req)
+                .expect(&format!("Failed to parse request in test case '{}'", $name)),
+            json_path: $jsonpaths,
+        }
+    };
+}
+
+lazy_static! {
+    /// Test case parameters (for [`run_test_case`]) to run for each model.
+    static ref TEST_CASES: Vec<TestCase> = vec![
+        test_case!("basic", json!({
+            "model": "not_needed",
+            "messages": [
+                {
+                    "role": "user",
+                    "content": "Say Hi"
+                }
+            ]
+        }), vec![
+            ("message_keys", "$.choices[*].message['role', 'tool_calls', 'refusal', 'function_calls']"),
+            ("replied_appropriately", "$.choices[*].message[?(@.content ~= 'Hi')].length()")
+        ]),
+    ];
+
+    /// Model instantiations to test.
+    static ref TEST_MODELS: Vec<(&'static str, Arc<dyn Chat>)> =
+        vec![("anthropic", anthropic::create_chat(None).expect("failed to create anthropic model"))];
+
+    /// A mapping of model names (in [`TEST_MODELS`]) and test names (in [`TEST_CASES`]) to skip.
+    static ref TEST_DENY_LIST: Vec<(&'static str, &'static str)> = vec![("anthropic", "advanced")];
+}
+
+/// Run a single [`TestCase`] for a model.
+#[allow(clippy::expect_used, clippy::expect_fun_call)]
+async fn run_test_case(
+    test: &TestCase,
+    model_name: &'static str,
+    model: Arc<dyn Chat>,
+) -> Result<(), anyhow::Error> {
+    let test_name = test.name;
+    println!("Running test {test_name}/{model_name} with {:?}", test.req);
+
+    let actual_resp = model
+        .chat_request(test.req.clone())
+        .await
+        .expect(format!("For test {test_name}/{model_name}, chat_request failed").as_str());
+
+    // Convert to [`serde_json::Value`] for JSONPath testing.
+    let resp_value = serde_json::to_value(&actual_resp).expect(
+        format!("For test {test_name}/{model_name}, failed to serialize response to JSON").as_str(),
+    );
+    for (id, json_ptr) in &test.json_path {
+        let resp_ptr = JsonPath::from_str(json_ptr)
+            .expect(format!("For test {test_name}, invalid JSONPath selector for id={id}").as_str())
+            .find(&resp_value);
+        insta::assert_snapshot!(
+            format!("{test_name}_{model_name}_{id}"),
+            serde_json::to_string_pretty(&resp_ptr).expect("Failed to serialize snapshot")
+        );
+    }
+    Ok(())
+}
+
+#[tokio::test]
+#[allow(clippy::expect_used, clippy::expect_fun_call)]
+async fn run_all_tests() {
```

Comment:
Should we allow running tests only for specific model - Anthropic, OpenAI, HF, Can we have similar approach but single test per model provider or some configuration how to control this?